Switch to SCSS: base files

skjnldsv opened the pull request:
This is the base scss files for cleanup purposes.
This is a basic css to scss conversion with a manual check using [scss-diff](https://www.npmjs.com/package/css-diff)

Commits with cleanups will follow. This is a simple pr.

**Please only use this pr to discuss issues or comments about its commits.
For arguing and/or ask question, go to the issue.**

References: #1786 & #1805 
@jancborchardt @rullzer @nickvergessen @LukasReschke @MorrisJobke @Henni @BernhardPosselt 
@nextcloud/designers

skjnldsv committed:
Cache and compile base

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Espina2 commented:
You are changing to scss?

skjnldsv commented:
> You are changing to scss?
Please see #1786

skjnldsv commented on core/css/multiselect.scss:
```diff
@@ -0,0 +1,125 @@
+		> input[type="checkbox"] + label:before {
+			margin-left: 7px;
+		}
+		input[type='checkbox'] {
```
Will do a cleanup after. :)

skjnldsv committed:
Debug log fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

Henni commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,115 @@
+		}
+
+		if(file_put_contents($this->fileCache, $this->rebaseUrls($compiledScss))) {
+			$this->logger->debug($root.'/'.$file.' compiled and successfully cached', ['app' => 'SCSSPHP']);
```
I believe the app should be set to 'core'.
Additionally it would probably make sense to log an error if `file_put_contents` fails.

skjnldsv replied:
It's going to change when I will implement appdata. But you're right, I should have done that :)

skjnldsv committed:
Appdata integration 1 & log fix 2

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Appdata integration 2

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

nickvergessen commented on lib/public/ICssManager.php:
```diff
@@ -0,0 +1,31 @@
+
+interface ICssManager {
+	
+	public function getCss($fileName);
```
missing `@since` on method and class

nickvergessen commented on core/Controller/CssController.php:
```diff
@@ -0,0 +1,79 @@
+ * @author Lukas Reschke <lukas@statuscode.ch>
+ * @author Morris Jobke <hey@morrisjobke.de>
+ *
+ * @license AGPL-3.0
```
This is a new file, right?
If so please replace the header

skjnldsv replied:
Oups! :)

nickvergessen commented on core/Controller/CssController.php:
```diff
@@ -0,0 +1,79 @@
+	protected $timeFactory;
+
+
+	public function __construct($appName, IRequest $request, ICssManager $cssManager, TimeFactory $timeFactory) {
```
Docs

skjnldsv replied:
Incoming

nickvergessen commented on lib/private/CssManager.php:
```diff
@@ -0,0 +1,58 @@
+		$this->appData = $appData;
+	}
+
+	public function getCss($fileName) {
```
docs

nickvergessen commented on lib/private/CssManager.php:
```diff
@@ -0,0 +1,58 @@
+		}
+
+	}
+}
```
new line

nickvergessen commented on lib/private/Template/CSSResourceLocator.php:
```diff
@@ -26,12 +26,22 @@
+
+	protected $appData;
+
+	public function __construct(\OCP\ILogger $logger, $theme, $core_map, $party_map, $appData) {
```
docs

nickvergessen commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,134 @@
+
+	/** @var \OCP\ILogger */
+	protected $logger;
+	protected $appData;
```
missing docs on almost all members

skjnldsv commented on core/Controller/CssController.php:
```diff
@@ -0,0 +1,79 @@
+		try {
+			$cssFile = $this->cssManager->getCss($fileName);
+		} catch(NotFoundException $e) {
+			return "not found";
```
Forgot that thanks

skjnldsv commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -108,27 +124,40 @@ private function cache() {
+	 */
 	public function getCachedSCSS() {
-		return $this->fileCache;
+		return "index.php/css/".$this->fileNameCSS;
```
We should probably use a URLGenerator?

skjnldsv commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -108,27 +124,40 @@ private function cache() {
 	private function rebaseUrls($css) {
 		$re = '/url\([\'"](.*)[\'"]\)/x';
-		$subst = 'url(\'../'.$this->rootCssLoc.'/$1\')';
+		$subst = 'url(\'../../../'.$this->rootCssLoc.'/$1\')';
```
I don't really have a better way to do that 😕

skjnldsv committed:
log app name fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
@nextcloud/designers please review

nickvergessen commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,163 @@
+	 * @return string
+	 */
+	public function getCachedSCSS() {
+		return "index.php/css/".$this->fileNameCSS;
```
`IURLGenerator::linkToRoute('core.Css.getCss')` should work

skjnldsv replied:
Thanks

skjnldsv committed:
Url generator fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Since fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
Rebased and ready to merge!

rullzer reviewed (changes requested):
Can you pleae crate a PR in the 3rdparty repo for this: https://github.com/nextcloud/3rdparty

rullzer commented on lib/private/Server.php:
```diff
@@ -406,6 +406,11 @@ public function __construct($webRoot, \OC\Config $config) {
 		});
+		$this->registerService('CssManager', function (Server $c) {
+			return new CssManager(
+				$c->getAppDataDir('server')
```
Please don't call this server. Call it 'css' or something. Server is way to generic.

rullzer commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,164 @@
+		} else {
+			return $this->cache();
+		}
+		return false;
```
This is never called.

rullzer commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,164 @@
+		} catch(NotFoundException $e) {
+			return false;
+		}
+        return false;
```
indentation is off

rullzer commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,164 @@
+	 * @return string
+	 */
+	public function getCachedSCSS() {
+		return substr(\OC::$server->getURLGenerator()->linkToRoute('core.Css.getCss', array('fileName' => $this->fileNameCSS)), 1);
```
Please inject the URLGenerator as well

rullzer commented on lib/private/TemplateLayout.php:
```diff
@@ -195,7 +195,8 @@ static public function findStylesheetFiles($styles) {
 			array( \OC::$SERVERROOT => \OC::$WEBROOT ),
-			array( \OC::$SERVERROOT => \OC::$WEBROOT ));
+			array( \OC::$SERVERROOT => \OC::$WEBROOT ),
+			\OC::$server->getAppDataDir('server'));
```
Same here. use something like `css`

skjnldsv replied:
So it will be css/css?

rullzer replied:
Thinking more about this. Will this stuff only be for core. Or are we planing on doing it also for apps? Because if so then it would make sense to name to folders after appid. 

@LukasReschke toughts?

skjnldsv replied:
@rullzer @LukasReschke We could do something like `appdata/css/appid`?

So for core `css/core` ?
It would be a shame not to allow apps to use the scss caching system :)

nickvergessen replied:
Well that should happen automatically right? if an app adds a scss file it gets added,if an old css file was used that one gets added...

skjnldsv replied:
Right now it's not like this. There's two types of css injection inside nextcloud:
1. Automatic search inside core/css
2. function OC\Util::add_style

The first one is only implemented right now.

rullzer commented on lib/public/ICssManager.php:
```diff
@@ -0,0 +1,38 @@
+<?php
```
Is this file really needed? I mean once it is in OCP we have to keep it stable for a long long time. This is not useful for other apps right?

skjnldsv replied:
Should I just remove it? I inspired myself from the avatar pr, so I figured this was needed! 😕

rullzer commented on lib/private/CssManager.php:
```diff
@@ -0,0 +1,65 @@
+ *
+ */
+
+namespace OC;
```
I think it would make sense to namespace this in `OC\CSS` or something. @nickvergessen suggetions?

rullzer commented on core/Controller/CssController.php:
```diff
@@ -0,0 +1,83 @@
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
```
nope! @LukasReschke

LukasReschke replied:
Set this to `@PublicPage` and `@NoCSRFRequired`. That's fine for this case since the data should be public. Not having PublicPage would make CSS unavailable for logged-in users.

rullzer replied:
Ah right. Forgot about public pages :P :see_no_evil:

rullzer commented on core/Controller/CssController.php:
```diff
@@ -0,0 +1,83 @@
+			$response->addHeader('Expires', $expires->format(\DateTime::RFC2822));
+			$response->addHeader('Pragma', 'cache');
+			return $response;
+		} else {
```
Drop the else.

rullzer commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,164 @@
+			$cachedfile->putContent($this->rebaseUrls($compiledScss));
+			$this->logger->debug($this->rootCssLoc.'/'.$this->fileNameSCSS.' compiled and successfully cached', ['app' => 'server']);
+			return true;
+		} catch(NotFoundException $e) {
```
putcontent doesn't throw NotFound

rullzer commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,164 @@
+	private function cache() {
+		$scss = new Compiler();
+		$scss->setImportPaths($this->fileLoc);
+		if(\OC::$server->getSystemConfig()->getValue('debug')) {
```
Inject IConfig

rullzer commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,164 @@
+
+class SCSSCacher {
+
+	protected $root;
```
Please add PHPdoc

rullzer commented on lib/private/Template/CSSResourceLocator.php:
```diff
@@ -26,12 +26,29 @@
 
 class CSSResourceLocator extends ResourceLocator {
+
+	protected $appData;
```
PHPdoc

rullzer commented on lib/private/CssManager.php:
```diff
@@ -0,0 +1,65 @@
+	public function getCss($fileName) {
+		try {
+			$folder = $this->appData->getFolder('css');
+		} catch(NotFoundException $e) {
```
No need to catch if you just throw again

rullzer commented on lib/private/CssManager.php:
```diff
@@ -0,0 +1,65 @@
+		}
+		try {
+			return $folder->getFile($fileName);
+		} catch(NotFoundException $e) {
```
No need to catch if you just throw again

skjnldsv replied:
We can't remove the catch in php try iirc. 😕

rullzer commented on lib/private/CssManager.php:
```diff
@@ -0,0 +1,65 @@
+	 * Get the css file and return ISimpleFile
+	 *
+	 * @param string $fileName css filename with extension
+	 * @return ISimpleFile
```
Missing `@throws`

LukasReschke commented on lib/private/Template/CSSResourceLocator.php:
```diff
@@ -26,12 +26,29 @@
+	 * @param array $party_map
+	 * @param IAppData $appData
+	 */
+	public function __construct(\OCP\ILogger $logger, $theme, $core_map, $party_map, $appData) {
```
$coreMap and $partyMap. Use camel-case instead of underscores :)

skjnldsv replied:
I used the same as ResourceLocator 😉

skjnldsv committed:
Injection fix and log appname fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Regex fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Appdata location fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Css cache folder name fix, route fix and various fixes

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Fix php docs

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

rullzer committed:
Code cleanup

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Do not add ICssManager to OCP

We can add it later if needed

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

skjnldsv commented:
Very nice @rullzer 👍 🎉 
I need to work on the tests and rebase, then we're good :)

skjnldsv commented:
Rebased and ready to review.

skjnldsv commented:
We need to re-add the ParseException error when a scss fail.
And we have an issue with the installation page. We can't use scss since appdata and database isn't initialised 😕

skjnldsv commented:
Script i'm using to convert css files to scss.
I was doing it manually at first. Don't forget to check for differences with  [scss-diff](https://www.npmjs.com/package/css-diff)

```
#!/bin/zsh

filename=$(basename "$1")
extension="${filename##*.}"
filename="${filename%.*}"
root=$(dirname "${1}")

scss="$root/$filename.scss"

echo "Converting $filename.$extension to $scss"

curl 'http://css2sass.herokuapp.com/' -H 'Pragma: no-cache' -H 'Origin: http://css2sass.herokuapp.com' -H 'Accept-Encoding: gzip, deflate' -H 'Accept-Language: en-us,en;q=0.5' -H 'Upgrade-Insecure-Requests: 1' -H 'Content-Type: application/x-www-form-urlencoded' -H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8' -H 'Cache-Control: no-cache' -H 'Referer: http://css2sass.herokuapp.com/' -H 'Connection: keep-alive' --data 'page%5Bcss%5D=&page%5Bsass%5D=&commit=Convert+2+SCSS' --data-urlencode "page[css]=`cat $1`" --compressed -s |  pup '#page_sass text{}' | recode html..ascii | sed 's/  /\t/g' | sed "s/\"/'/g" > $scss
```

`find -name "*.css" -exec css2scss {} \;`

skjnldsv committed:
Css installation page fallback

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
> <skjnldsv> On a fresh nc installation
> <skjnldsv> without things set up
> <skjnldsv> The scss appdata can't work since it's using the database
> <skjnldsv> and nothing is set
> <skjnldsv> How do you suggest we serve css files in this particular case?
> <skjnldsv> What I've thought so far:
> <skjnldsv> 1. Creating a install.css file for this particular case only
> <skjnldsv> 2. serving the compiled css on-the-fly without caching until database settings are ok
> <skjnldsv> Any ideas?

I'm thinking going for the installation.css file. Much easier. But we should keep as simple as we can.
If you got any counter-arguments please say. 😄 

@jancborchardt @LukasReschke @nextcloud/designers

skjnldsv committed:
Installation css & template edit

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

jancborchardt commented:
@skjnldsv we need to wait with merging until we branched off stable11 anyway. :) Then we can get it in directly at the beginning of the Nextcloud 12 cycle when everything is fine

skjnldsv commented:
@jancborchardt You're talking about the css files?

jancborchardt commented:
@skjnldsv I mean this pull request :) Since we’re in feature freeze for Nextcloud 11 which will be released soon, this will make it into Nextcloud 12. :)

skjnldsv commented:
Good! :)
There's still some work to do. Tests for example. I'd like to make some.

rullzer commented:
@skjnldsv so after having another look (all that changed CSS does not make it to easy). I do have some more questions.

* Why the CSSManager? That seems to just be a dump wrapper around the AppData. No need for that if it stays the way it is now.

* So you compile each css file into the scss file right? This is done when creating a template and not when requesting a stylesheet (right?). Wouldn't it be simple to compile ondemand when it is requested from the controller?

skjnldsv commented:
> Why the CSSManager? That seems to just be a dump wrapper around the AppData. No need for that if it stays the way it is now.

Because I used the overall file hierarchy nectcloud seems to use for other components (avatars...)
I figured it would be more pleasant not to change everything!

> So you compile each css file into the scss file right? This is done when creating a template and not when requesting a stylesheet (right?). Wouldn't it be simple to compile ondemand when it is requested from the controller?

This is indeed done on the first template creation. Having the compilation don on request would change the entire way nextcloud is handling css files. Like above, I dodn't wanted to change everything, and it seemed much simpler and easy to understand to use something like I did than creating an entire other system. Of course I could have created a OC\CSS part and create an independent solution, but since this is already implemented on the template generation, I kind of liked it this way ;)

skjnldsv commented on lib/private/Template/ResourceLocator.php:
```diff
@@ -107,6 +107,38 @@ protected function appendIfExist($root, $file, $webRoot = null) {
+	 * @param string|null $webRoot base for path, default map $root to $webRoot
+	 * @return bool True if the resource was found and cached, false otherwise
+	 */
+	protected function cacheAndAppendScssIfExist($root, $file, $appData, $urlGenerator, $systemConfig, $webRoot = null) {
```
Like I said on the other reply, didn't wanted to change everything 😂

rullzer committed:
No need for the CssManager

* It is a simple wrapper we can always add it later if needed

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Formatting

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Make SCCCacher injectable

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Inject SCSSCacher

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Fix typo

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

skjnldsv committed:
Apps scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Header scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Icons scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Inputs scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Jquery-ui-fixes scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Multiselect scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Share scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Styles scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Systemtags scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Tooltip scss

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
3rdparty branch set

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
Removed old css files

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
Okay, I dropped old scss converted files, and updated the new ones!
This is now or never! :)

I'm going to fix some stuff on the installation css file, but feel free to test everything asap.
Every commit on master will be a pain in the a** to fix again because of this conversion.

skjnldsv committed:
Updated installation css for config error and update

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MorrisJobke commented on lib/private/Template/CSSResourceLocator.php:
```diff
@@ -25,13 +25,34 @@
 
 namespace OC\Template;
```
Please drop the executable bit on this file ;) `chmod -x lib/private/Template/CSSResourceLocator.php`

skjnldsv replied:
I will to a quick permission fix yes

LukasReschke replied:
> Please drop the executable bit on this file ;) chmod -x lib/private/Template/CSSResourceLocator.php
  
Time for a CI check for this? :)

MorrisJobke commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,183 @@
+<?php
```
Same for `lib/private/Template/ResourceLocator.php` - remove the executable bit.

MorrisJobke commented:
If I check out this branch I get following in the log file:

```
{"reqId":"j9WVXoY7TSKDwU1qSMUa","remoteAddr":"::1","app":"PHP","message":"Only variables should be passed by reference at \/Users\/morrisjobke\/Projects\/nextcloud\/server\/lib\/private\/Template\/SCSSCacher.php#178","level":0,"time":"2016-12-22T14:13:02+00:00","method":"GET","url":"\/server\/index.php\/apps\/files\/?dir=\/","user":"admin","version":"12.0.0.11"}
```


And those requests return a 404:

![bildschirmfoto 2016-12-22 um 15 13 18](https://cloud.githubusercontent.com/assets/245432/21428223/4135c3ca-c859-11e6-9412-2d01f7d79639.png)

skjnldsv committed:
Fix variables should be passed by reference

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MorrisJobke commented:
With short URLs (index.php removed):

![bildschirmfoto 2016-12-22 um 15 43 12](https://cloud.githubusercontent.com/assets/245432/21429317/8da1fdaa-c85e-11e6-9890-d9ace29477a2.png)
![bildschirmfoto 2016-12-22 um 15 45 24](https://cloud.githubusercontent.com/assets/245432/21429319/8daa7354-c85e-11e6-8a8c-634aabdcfd54.png)

Without short URLs:

![bildschirmfoto 2016-12-22 um 15 48 34](https://cloud.githubusercontent.com/assets/245432/21429324/94d19c84-c85e-11e6-90fa-c83ffacc6902.png)

skjnldsv commented:
@rullzer this seems related to the urlgenerator and the url rewriting of apache. Would you mind taking a quick look? :D

MorrisJobke commented:
> @rullzer this seems related to the urlgenerator and the url rewriting of apache. Would you mind taking a quick look? :D

Ping @rullzer

rullzer commented:
Somehow I missed this. This is high on my todo for this week! :)

rullzer commented:
ah apache stuff... always fun. let me setup that.

rullzer commented:
@MorrisJobke I can't reproduce your issue. Do you have more detailed logs for me?

ChristophWurst commented on lib/private/Template/SCSSCacher.php:
```diff
@@ -0,0 +1,184 @@
+	 * @param string $path
+	 * @return boolean
+	 */
+	private function is_cached($fileNameCSS, $fileNameSCSS, ISimpleFolder $folder, $path) {
```
please rename to `isCached`

rullzer committed:
Adds CssControllerTests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

LukasReschke commented:
> @MorrisJobke I can't reproduce your issue. Do you have more detailed logs for me?

I can reproduce this. Let's see…

LukasReschke commented on lib/private/Template/CSSResourceLocator.php:
```diff
@@ -25,13 +25,34 @@
+	 * @param string $theme
+	 * @param array $core_map
+	 * @param array $party_map
+	 * @param SCSSCacher $scssCcacher
```
Typo

LukasReschke committed:
Remove webroot from file path

The webroot is already provided, otherwise this links to stuff like "/nextcloud/nextcloud/index.php" instead of "/nextcloud/index.php"

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke commented:
https://github.com/nextcloud/server/pull/2050/commits/b2d7f75ead9043cc5a7203fbe42a9925216ce281 should fix @MorrisJobke's issue.

However, the update interface is pretty much broken. I suppose we need some hard-coded CSS there?

skjnldsv committed:
Camelcase fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv committed:
ScssCcacher typo fix

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

MorrisJobke commented:
> However, the update interface is pretty much broken. I suppose we need some hard-coded CSS there?

Yes - this is because executing PHP during an upgrade is not permitted and the "this instance is currently in update mode" message is shown instead of the CSS. So we need some static CSS for this page. But this shouldn't be that hard.

skjnldsv commented:
Yeah, I was merging the properties into one, but didn't finished. I will take a look.

MorrisJobke commented:
Everything works except the icons. The icons.css contains a `../` too much and therefore doesn't take care of the subfolder my instance is installed on (it's installed in `localhost/server/` and the icons.css refers to `localhost/core` instead of `localhost/server/core`). @skjnldsv Could you check that one?

skjnldsv commented:
@MorrisJobke can't reproduce.
The page `http://localhost:8088/server/apps/files/` load the `http://localhost:8088/server/index.php/css/core/icons.css` file which contains relative paths like this one: `.icon-download{background-image:url('../../../core/css/../img/actions/download.svg?v=1')}`

Are the css files loaded with url rewriting too on your side? That would explain the issue, but this is strange since css files are supposed to be excluded from url rewrite right?

skjnldsv commented:
This is strange because on my nginx conf, the regular css files are requested without the index.php, but the scss generated css files are requested with the index.php, so it shouldn't be an issue. But I will work on it, since we know where it comes from.

MorrisJobke commented:
> The page http://localhost:8088/server/apps/files/ load the http://localhost:8088/server/index.php/css/core/icons.css file which contains relative paths like this one: .icon-download{background-image:url('../../../core/css/../img/actions/download.svg?v=1')}

I don't have an `index.php` in there - maybe this is the problem then. 😕

skjnldsv commented:
yep, this is the issue. I need to find a different way to calculate the different path.

skjnldsv committed:
Fix css url prefix for IgnoreFrontController enabled configs

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
@MorrisJobke please test again

rullzer committed:
Update autoloader

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
* rebased 3rdparty
* rebased this on master

MorrisJobke commented:
JSUnit tests fail:

```
PhantomJS 2.1.1 (Linux 0.0.0) OCA.Files.BreadCrumb tests Resizing Hides breadcrumbs to fit max allowed width FAILED
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:210:52
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:211:52
PhantomJS 2.1.1 (Linux 0.0.0) OCA.Files.BreadCrumb tests Resizing Updates the breadcrumbs when reducing max allowed width FAILED
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:239:52
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:240:52
	Expected true to equal false.
	apps/files/tests/js/breadcrumbSpec.js:241:52
PhantomJS 2.1.1 (Linux 0.0.0) OCA.Trashbin.FileList tests Breadcrumbs links the breadcrumb to the trashbin view FAILED
	Expected 3 to equal 2.
	apps/files_trashbin/tests/js/filelistSpec.js:135:34
	Expected '...' to equal 'subdir'.
	apps/files_trashbin/tests/js/filelistSpec.js:139:50
	Expected 'http://localhost/index.php/apps/files/?dir=/subdir' to equal 'http://localhost/index.php/apps/files?view=trashbin&dir=/subdir'.
	apps/files_trashbin/tests/js/filelistSpec.js:141:13
```

MorrisJobke commented:
> I don't have an index.php in there - maybe this is the problem then. 😕

Yes 👍 Works now.

Now only the problem with the update page is there.

MorrisJobke commented:
> JSUnit tests fail:

Let's ignore them for now - seem to be not related (maybe they are broken on master as well)

MorrisJobke commented:
The list of missing files for the update page:

```
header.css
styles.css
apps.css
icons.css
multiselect.css
jquery-ui-fixes.css
tooltip.css
inputs.css
```

Let me check what I could reuse here.

MorrisJobke committed:
Theme update pages via CSS

* SCSS on-the-fly generation isn't allowed during update
* fallback to plain CSS

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
I fixed the update page, the "this instance is in maintenance mode" page and the "this instance is too big" page with the CSS rules from the missing CSS files combined and shrinked into the update.css.

MorrisJobke reviewed (approved):
Tested and works here 👍

skjnldsv commented:
@MorrisJobke That's why I created the installation.css file. Maybe we should merge the two of them?

rullzer committed:
Fix js tests

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

MorrisJobke commented:
> @MorrisJobke That's why I created the installation.css file. Maybe we should merge the two of them?

I would keep them separate, because they don't share that much rules. Most of the rules of installation are regarding input fields and for the update page it's mostly display status messages and info texts.

rullzer commented:
Lets do this! :rocket:

skjnldsv commented:
Youhouuuuuuu

LukasReschke commented:
Please fix the install page ASAP or I will revert this. Having the following in master is not an option:


![screen shot 2017-01-09 at 12 29 12](https://cloud.githubusercontent.com/assets/878997/21765151/5749cabe-d667-11e6-8d2c-52545c3746a0.png)

LukasReschke commented:
@MorrisJobke told me he is on it. I suppose the CSS links are generated wrongly:

```
					<link rel="stylesheet" href="/core/css/installation.css?v=ba222ded25d957b900c03bef914333cd">
```

My Nc is installed under a subfolder.

MorrisJobke commented:
> My Nc is installed under a subfolder.

Same here.

MorrisJobke commented:
> Please fix the install page ASAP or I will revert this. Having the following in master is not an option:

Fixed with #2992
